demos/projects/ESPRESSIF/az-ca-recovery/components/sample-azure-iot-ca-recovery/azure_trust_bundle_storage_esp.c

```diff
@@ -0,0 +1,185 @@
+    esp_err_t err;
+    size_t ulTrustBundleReadSize = 0; /* value will default to 0, if not set yet in NVS */
+    size_t ulTrustBundleVersionReadSize = 0;
+
+    err = nvs_open( CA_CERT_NAMESPACE, NVS_READWRITE, &xNVSHandle );
+
+    if( err != ESP_OK )
+    {
+        ESP_LOGE( TAG, "Error (%s) opening NVS!\n", esp_err_to_name( err ) );
+        return eAzureIoTErrorFailed;
+    }
+
+    err = nvs_get_blob( xNVSHandle, AZURE_TRUST_BUNDLE_VERSION_NAME, NULL, &ulTrustBundleVersionReadSize );
+
+    if( err != ESP_OK )
+    {
+        ESP_LOGE( TAG, "Error (%s) getting AZURE_TRUST_BUNDLE_VERSION_NAME from NVS!\n", esp_err_to_name( err ) );
+        nvs_close( xNVSHandle );
+        return eAzureIoTErrorFailed;
+    }
+
+    if( ulTrustBundleVersionReadSize > ulTrustBundleVersionLength )
+    {
+        ESP_LOGE( TAG, "Not enough space to read version\n" );
+        nvs_close( xNVSHandle );
+        return eAzureIoTErrorOutOfMemory;
+    }
+
+    *pulOutTrustBundleVersionLength = ulTrustBundleVersionReadSize;
+
+    err = nvs_get_blob( xNVSHandle, AZURE_TRUST_BUNDLE_VERSION_NAME, &pucTrustBundleVersion, pulOutTrustBundleVersionLength );
+
+    if( err != ESP_OK )
+    {
+        nvs_close( xNVSHandle );
+        return eAzureIoTErrorFailed;
+    }
+
+    /* Get size of cert */
+    err = nvs_get_blob( xNVSHandle, AZURE_TRUST_BUNDLE_NAME, NULL, &ulTrustBundleReadSize );
+
+    if( ( err != ESP_OK ) && ( err != ESP_ERR_NVS_NOT_FOUND ) )
+    {
+        nvs_close( xNVSHandle );
+        return eAzureIoTErrorFailed;
+    }
+
+    if( ulTrustBundleReadSize == 0 )
+    {
+        ESP_LOGE( TAG, "Nothing saved yet!\n" );
+    }
+    else if( ulTrustBundleReadSize > ulTrustBundleLength )
+    {
+        ESP_LOGE( TAG, "Not enough space to read trust bundle" );
+    }
+    else
+    {
+        *pulOutTrustBundleLength = ulTrustBundleReadSize;
+        err = nvs_get_blob( xNVSHandle, AZURE_TRUST_BUNDLE_NAME, ( void * ) pucTrustBundle, pulOutTrustBundleLength );
+
+        if( err != ESP_OK )
+        {
+            nvs_close( xNVSHandle );
+            return eAzureIoTErrorFailed;
+        }
+    }
+
+    nvs_close( xNVSHandle );
+
+    return eAzureIoTSuccess;
+}
+
+AzureIoTResult_t AzureIoTCAStorage_WriteTrustBundle( const uint8_t * pucTrustBundle,
+                                                     uint32_t ulTrustBundleLength,
+                                                     const uint8_t * pucTrustBundleVersion,
+                                                     uint32_t ulTrustBundleVersionLength )
+{
+    nvs_handle_t xNVSHandle;
+    esp_err_t err;
+    size_t ulTrustBundleVersionReadSize = 0;
+
+    err = nvs_open( CA_CERT_NAMESPACE, NVS_READWRITE, &xNVSHandle );
+
+    if( err != ESP_OK )
+    {
+        ESP_LOGE( TAG, "Error (%s) opening NVS!\n", esp_err_to_name( err ) );
+        return eAzureIoTErrorFailed;
+    }
+
+    uint8_t ucReadTrustBundleVersion[ 8 ] = { 0 }; /* value will default to 0, if not set yet in NVS */
```

Comment:
Version should be always written last. If the version is written first and the file write fails/ends up corrupted, next time the device boots, it will refuse to re-download the trustbundle and recover because it has the same version with the one stored, even if the actual contents is corrupted.

Version write is in other words synonymous with transaction commit in this case.

Reply:
Yup it is. This section is only for fetching the version to compare.